iceoryx2/src/service/attribute.rs

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Attributes can be defined for a [`crate::service::Service`]. They define features that do not
+//! change during the lifetime of the [`crate::service::Service`] and are accessible by anyone that
+//! is allowed to open the [`crate::service::Service`].
+//!
+//! ## Create Service With Attributes
+//!
+//! ```
+//! use iceoryx2::prelude::*;
+//! use iceoryx2::config::Config;
+//!
+//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! let service_name = ServiceName::new("My/Funk/ServiceName")?;
+//!
+//! let service_creator = zero_copy::Service::new(&service_name)
+//!     .publish_subscribe::<u64>()
+//!     .create_with_attributes(
+//!         // all attributes that are defined when creating a new service are stored in the
+//!         // static config of the service
+//!         &DefinedAttributes::new()
+//!             .define("some attribute key", "some attribute value")
+//!             .define("some attribute key", "another attribute value for the same key")
+//!             .define("another key", "another value")
+//!     )?;
+//!
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! ## Open Service With Attribute Requirements
+//!
+//! ```no_run
+//! use iceoryx2::prelude::*;
+//! use iceoryx2::config::Config;
+//!
+//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! let service_name = ServiceName::new("My/Funk/ServiceName")?;
+//!
+//! let service_open = zero_copy::Service::new(&service_name)
+//!     .publish_subscribe::<u64>()
+//!     .open_with_attributes(
+//!         // All attributes that are defined when opening a new service interpreted as
+//!         // requirements.
+//!         // If a attribute key as either a different value or is not set at all, the service
+//!         // cannot be opened. If not specific attributes are required one can skip them completely.
+//!         &RequiredAttributes::new()
+//!             .require("another key", "another value")
```

Comment:
I also thought of this but with the `AttributeSet` (the type which is returned when the user just wants to know what attributes are defined) and the `DefinedAttributes` it makes somehow sense, since this type shall only have `require()` as method. With the `Attributes` type one may uses this also for something else, for instance to store attributes and this is something it shall not do.

* `optional` is not necessary, if you have an optional requirement you do not need to specify it at all.
* instead of `present` what do you think of `require_key`? I like the word `require` in this context since it really screams at you that you require **. Present does not have the same notion.

Reply:
The repetition of `require` is somewhat annoying. What do you think of `AttributeVerifier::require` and `AttributeVerifier::require_key`?